File: client/typescript/src/world/debugger.ts
import type { MemoryRange } from "../_generated/memory/address/map.js";
import type { BreakpointId } from "../_generated/program/outcome.js";
import type { WatchpointId } from "../_generated/program/watch.js";
import type { Breakpoint } from "../_generated/runtime/debugger/breakpoint.js";
import type {
    EvaluationLimits,
    EvaluationMode,
    EvaluationOutcome,
    EvaluationRepresentation,
    EvaluationTarget,
    EvaluationValue,
} from "../_generated/runtime/debugger/evaluation.js";
import type {
    Execution,
    Step,
} from "../_generated/runtime/debugger/execution.js";
import type * as debugFrame from "../_generated/runtime/debugger/frame.js";
import type {
    Allocation,
    AllocationId,
    Reference,
    Root,
    RootId,
    RootSource,
} from "../_generated/runtime/debugger/heap.js";
import type {
    MemoryMap,
    Region,
    RegionKind,
} from "../_generated/runtime/debugger/memory.js";
import type {
    Probe,
    ProbeAction,
    ProbeFilter,
    ProbeId,
} from "../_generated/runtime/debugger/probe.js";
import type { Watchpoint } from "../_generated/runtime/debugger/watchpoint.js";
import type { RuntimeId } from "../_generated/runtime/runtime.js";
import type * as debuggerService from "../_generated/runtime/service/debugger.js";
import type { WorkerId } from "../_generated/runtime/worker/worker.js";
import type { Moment } from "../_generated/runtime/world/lineage/moment.js";
import type { RunOutcome } from "../_generated/runtime/world/world/run.js";
import { DebuggerClient } from "../_generated/world/debugger.js";

import type { EvaluationOptions } from "./runtime.js";
import type { World } from "./world.js";

/** Frame query fields bound to one World. */
export type FrameOptions = Omit<debuggerService.ReadFramesRequest, "worldId">;

/** One World-bound Destack debugger. */
export class Debugger {
    /** Owning World. */
    readonly world: World;
    /** Complete generated Debugger service client. */
    readonly client: DebuggerClient;

    /** Bind one Debugger to its World. */
    constructor(world: World) {
        this.world = world;
        this.client = new DebuggerClient(world.connection);
    }

    // =============================================================================
    // Execution
    // =============================================================================

    /** Pause selected execution in this World. */
    async pause(execution: Execution = { kind: "world" }): Promise<void> {
        await this.client.pause({ worldId: this.world.id, execution });
    }

    /** Resume one debugger-stopped Worker. */
    async resume(runtimeId: RuntimeId, workerId: WorkerId): Promise<RunOutcome> {
        const response = await this.client.resume({
            worldId: this.world.id,
            runtimeId,
            workerId,
        });

        return response.value;
    }

    /** Step one debugger-stopped Frame. */
    async step(
        frameId: debugFrame.FrameId,
        step: Step,
    ): Promise<RunOutcome> {
        const response = await this.client.step({
            worldId: this.world.id,
            frameId,
            step,
        });

        return response.value;
    }

    // =============================================================================
    // Evaluation
    // =============================================================================

    /** Evaluate one TS++ expression in a selected execution context. */
    async evaluate(
        target: EvaluationTarget,
        expression: string,
        options: EvaluationOptions,
    ): Promise<debuggerService.Evaluation> {
        const response = await this.client.evaluate({
            worldId: this.world.id,
            target,
            expression,
            ...options,
        });

        return response.value;
    }

    // =============================================================================
    // Frame
    // =============================================================================

    /** Read captured Frames from this World. */
    async frames(options: FrameOptions = {}): Promise<Frame[]> {
        const response = await this.client.readFrames({
            worldId: this.world.id,
            ...options,
        });

        return response.value.map((frame) => new Frame(this.world, frame));
    }

    // =============================================================================
    // Memory
    // =============================================================================

    /** Read this World's mapped memory Regions. */
    async memoryMap(
        moment?: debuggerService.ReadMemoryMapRequest["moment"],
    ): Promise<MemoryMap> {
        const response = await this.client.readMemoryMap({
            worldId: this.world.id,
            moment,
        });

        return response.value;
    }

    /** Stream one exact memory range from this World. */
    async *memory(
        moment: Moment,
        range: MemoryRange,
    ): AsyncGenerator<Uint8Array> {
        const call = this.client.readMemory({
            worldId: this.world.id,
            moment,
            range,
        });

        // normalize every transport byte batch
        for await (const bytes of call) {
            yield bytes instanceof Uint8Array ? bytes : Uint8Array.from(bytes);
        }

        await call.response();
    }

    // =============================================================================
    // Heap
    // =============================================================================

    /** Stream managed heap Allocations at one exact Moment. */
    async *allocations(
        moment: Moment,
        after?: AllocationId,
    ): AsyncGenerator<Allocation> {
        const call = this.client.listAllocations({
            worldId: this.world.id,
            moment,
            after,
        });

        // flatten transport batches into semantic Allocations
        for await (const allocations of call) {
            for (const allocation of allocations) {
                yield allocation;
            }
        }

        await call.response();
    }

    /** Stream heap Roots at one exact Moment. */
    async *roots(moment: Moment, after?: RootId): AsyncGenerator<Root> {
        const call = this.client.listRoots({
            worldId: this.world.id,
            moment,
            after,
        });

        // flatten transport batches into semantic Roots
        for await (const roots of call) {
            for (const root of roots) {
                yield root;
            }
        }

        await call.response();
    }

    /** Stream outgoing References from one Allocation at one exact Moment. */
    async *references(
        moment: Moment,
        allocationId: AllocationId,
        after?: Reference["offset"],
    ): AsyncGenerator<Reference> {
        const call = this.client.listReferences({
            worldId: this.world.id,
            moment,
            allocationId,
            after,
        });

        // flatten transport batches into semantic References
        for await (const references of call) {
            for (const reference of references) {
                yield reference;
            }
        }

        await call.response();
    }

    // =============================================================================
    // Breakpoint
    // =============================================================================

    /** Return the Breakpoints installed in this World. */
    async breakpoints(
        moment?: debuggerService.ListBreakpointsRequest["moment"],
    ): Promise<readonly Breakpoint[]> {
        const response = await this.client.listBreakpoints({
            worldId: this.world.id,
            moment,
        });

        return response.value;
    }

    /** Add one Breakpoint to this World. */
    async addBreakpoint(filter: Breakpoint["filter"]): Promise<BreakpointId> {
        const response = await this.client.addBreakpoint({
            worldId: this.world.id,
            filter,
        });

        return response.value;
    }

    /** Replace one Breakpoint in this World. */
    async updateBreakpoint(breakpoint: Breakpoint): Promise<void> {
        await this.client.updateBreakpoint({
            worldId: this.world.id,
            breakpoint,
        });
    }

    /** Remove one Breakpoint from this World. */
    async removeBreakpoint(breakpointId: BreakpointId): Promise<void> {
        await this.client.removeBreakpoint({
            worldId: this.world.id,
            breakpointId,
        });
    }

    // =============================================================================
    // Watchpoint
    // =============================================================================

    /** Return the Watchpoints installed in this World. */
    async watchpoints(
        moment?: debuggerService.ListWatchpointsRequest["moment"],
    ): Promise<readonly Watchpoint[]> {
        const response = await this.client.listWatchpoints({
            worldId: this.world.id,
            moment,
        });

        return response.value;
    }

    /** Add one Watchpoint to this World. */
    async addWatchpoint(filter: Watchpoint["filter"]): Promise<WatchpointId> {
        const response = await this.client.addWatchpoint({
            worldId: this.world.id,
            filter,
        });

        return response.value;
    }

    /** Replace one Watchpoint in this World. */
    async updateWatchpoint(watchpoint: Watchpoint): Promise<void> {
        await this.client.updateWatchpoint({
            worldId: this.world.id,
            watchpoint,
        });
    }

    /** Remove one Watchpoint from this World. */
    async removeWatchpoint(watchpointId: WatchpointId): Promise<void> {
        await this.client.removeWatchpoint({
            worldId: this.world.id,
            watchpointId,
        });
    }

    // =============================================================================
    // Probe
    // =============================================================================

    /** Return the Probes installed in this World. */
    async probes(
        moment?: debuggerService.ListProbesRequest["moment"],
    ): Promise<readonly Probe[]> {
        const response = await this.client.listProbes({
            worldId: this.world.id,
            moment,
        });

        return response.value;
    }

    /** Add one Probe to this World. */
    async addProbe(filter: ProbeFilter, action: ProbeAction): Promise<ProbeId> {
        const response = await this.client.addProbe({
            worldId: this.world.id,
            filter,
            action,
        });

        return response.value;
    }

    /** Replace one Probe in this World. */
    async updateProbe(probe: Probe): Promise<void> {
        await this.client.updateProbe({ worldId: this.world.id, probe });
    }

    /** Remove one Probe from this World. */
    async removeProbe(probeId: ProbeId): Promise<void> {
        await this.client.removeProbe({ worldId: this.world.id, probeId });
    }
}

/** One captured execution Frame. */
export class Frame {
    /** World containing this Frame. */
    readonly world: World;
    /** Complete captured Frame record. */
    readonly record: debugFrame.Frame;

    /** Bind one captured Frame record to its World. */
    constructor(world: World, record: debugFrame.Frame) {
        this.world = world;
        this.record = record;
    }

    /** Frame identity. */
    get id(): debugFrame.FrameId {
        return this.record.id;
    }

    /** Canonical captured Frame bytes. */
    get bytes(): Uint8Array {
        return this.record.bytes instanceof Uint8Array
            ? this.record.bytes
            : Uint8Array.from(this.record.bytes);
    }

    /** Step this debugger-stopped Frame. */
    step(step: Step): Promise<RunOutcome> {
        return this.world.debugger.step(this.id, step);
    }

    /** Evaluate one TS++ expression in this Frame. */
    evaluate(
        expression: string,
        options: EvaluationOptions,
    ): Promise<debuggerService.Evaluation> {
        return this.world.debugger.evaluate(
            { kind: "frame", frame: this.id },
            expression,
            options,
        );
    }
}

export type {
    Allocation,
    AllocationId,
    Breakpoint,
    BreakpointId,
    EvaluationLimits,
    EvaluationMode,
    EvaluationOutcome,
    EvaluationRepresentation,
    EvaluationTarget,
    EvaluationValue,
    Execution,
    MemoryMap,
    MemoryRange,
    Probe,
    ProbeAction,
    ProbeFilter,
    ProbeId,
    Reference,
    Region,
    RegionKind,
    Root,
    RootId,
    RootSource,
    Step,
    Watchpoint,
    WatchpointId,
};
export { DebuggerClient };
export type { Evaluation } from "../_generated/runtime/service/debugger.js";
export { debuggerService } from "../_generated/world/debugger.js";

File: language/runtime/src/service/debugger.rs
use serde::{Deserialize, Serialize};
use tspp_memory::MemoryRange;
use tspp_program as program;
use tspp_rpc::service;
use tspp_serde::Reflect;

use crate::debugger::{
    Allocation, AllocationId, Breakpoint, EvaluationLimits, EvaluationMode, EvaluationOutcome,
    EvaluationTarget, Execution, Frame, FrameId, MemoryFilter, MemoryMap, PointFilter, Probe,
    ProbeAction, ProbeFilter, ProbeId, Reference, Root, RootId, Step, Watchpoint,
};
use crate::runtime::RuntimeId;
use crate::worker::WorkerId;
use crate::world::{Moment, RunOutcome};

use super::WorldId;

/// RPC operations over hosted World debuggers.
#[service(name = "tspp.world.Debugger")]
pub trait DebuggerService {
    // =============================================================================
    // Execution
    // =============================================================================

    /// Pause selected World execution.
    #[rpc(name = "Pause")]
    fn pause(request: PauseRequest) -> ();

    /// Resume one debugger-stopped Worker.
    #[rpc(name = "Resume")]
    fn resume(request: ResumeRequest) -> RunOutcome;

    /// Step one debugger-stopped Frame.
    #[rpc(name = "Step")]
    fn step(request: StepRequest) -> RunOutcome;

    // =============================================================================
    // Evaluation
    // =============================================================================

    /// Evaluate one TS++ expression in a selected execution context.
    #[rpc(name = "Evaluate")]
    fn evaluate(request: EvaluateRequest) -> Evaluation;

    // =============================================================================
    // Frame
    // =============================================================================

    /// Read captured Frames from one World.
    #[rpc(name = "ReadFrames", idempotency = "no_side_effects")]
    fn read_frames(request: ReadFramesRequest) -> Vec<Frame>;

    // =============================================================================
    // Memory
    // =============================================================================

    /// Read one World's mapped memory Regions.
    #[rpc(name = "ReadMemoryMap", idempotency = "no_side_effects")]
    fn read_memory_map(request: ReadMemoryMapRequest) -> MemoryMap;

    /// Stream one exact World memory range.
    #[rpc(name = "ReadMemory", response_stream(Vec<u8>), idempotency = "no_side_effects")]
    fn read_memory(request: ReadMemoryRequest) -> ();

    // =============================================================================
    // Heap
    // =============================================================================

    /// Stream managed heap Allocations in one World.
    #[rpc(
        name = "ListAllocations",
        response_stream(Vec<Allocation>),
        idempotency = "no_side_effects"
    )]
    fn list_allocations(request: ListAllocationsRequest) -> ();

    /// Stream heap Roots in one World.
    #[rpc(
        name = "ListRoots",
        response_stream(Vec<Root>),
        idempotency = "no_side_effects"
    )]
    fn list_roots(request: ListRootsRequest) -> ();

    /// Stream outgoing References from one heap Allocation.
    #[rpc(
        name = "ListReferences",
        response_stream(Vec<Reference>),
        idempotency = "no_side_effects"
    )]
    fn list_references(request: ListReferencesRequest) -> ();

    // =============================================================================
    // Breakpoint
    // =============================================================================

    /// List the Breakpoints in one World.
    #[rpc(name = "ListBreakpoints", idempotency = "no_side_effects")]
    fn list_breakpoints(request: ListBreakpointsRequest) -> Vec<Breakpoint>;

    /// Add one Breakpoint to a World.
    #[rpc(name = "AddBreakpoint")]
    fn add_breakpoint(request: AddBreakpointRequest) -> program::BreakpointId;

    /// Update one Breakpoint in a World.
    #[rpc(name = "UpdateBreakpoint")]
    fn update_breakpoint(request: UpdateBreakpointRequest) -> ();

    /// Remove one Breakpoint from a World.
    #[rpc(name = "RemoveBreakpoint", idempotency = "idempotent")]
    fn remove_breakpoint(request: RemoveBreakpointRequest) -> ();

    // =============================================================================
    // Watchpoint
    // =============================================================================

    /// List the Watchpoints in one World.
    #[rpc(name = "ListWatchpoints", idempotency = "no_side_effects")]
    fn list_watchpoints(request: ListWatchpointsRequest) -> Vec<Watchpoint>;

    /// Add one Watchpoint to a World.
    #[rpc(name = "AddWatchpoint")]
    fn add_watchpoint(request: AddWatchpointRequest) -> program::WatchpointId;

    /// Update one Watchpoint in a World.
    #[rpc(name = "UpdateWatchpoint")]
    fn update_watchpoint(request: UpdateWatchpointRequest) -> ();

    /// Remove one Watchpoint from a World.
    #[rpc(name = "RemoveWatchpoint", idempotency = "idempotent")]
    fn remove_watchpoint(request: RemoveWatchpointRequest) -> ();

    // =============================================================================
    // Probe
    // =============================================================================

    /// List the Probes in one World.
    #[rpc(name = "ListProbes", idempotency = "no_side_effects")]
    fn list_probes(request: ListProbesRequest) -> Vec<Probe>;

    /// Add one Probe to a World.
    #[rpc(name = "AddProbe")]
    fn add_probe(request: AddProbeRequest) -> ProbeId;

    /// Update one Probe in a World.
    #[rpc(name = "UpdateProbe")]
    fn update_probe(request: UpdateProbeRequest) -> ();

    /// Remove one Probe from a World.
    #[rpc(name = "RemoveProbe", idempotency = "idempotent")]
    fn remove_probe(request: RemoveProbeRequest) -> ();
}

/// Result of one hosted expression evaluation.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Reflect)]
pub struct Evaluation {
    /// World that performed the evaluation.
    pub world_id: WorldId,
    /// Moment containing the evaluation outcome.
    pub moment: Moment,
    /// Evaluation outcome.
    pub outcome: EvaluationOutcome,
}

/// Request to pause selected World execution.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct PauseRequest {
    /// World containing the selected execution.
    pub world_id: WorldId,
    /// Execution to pause.
    pub execution: Execution,
}

/// Request to resume one debugger-stopped Worker.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct ResumeRequest {
    /// World containing the stopped Worker.
    pub world_id: WorldId,
    /// Runtime containing the stopped Worker.
    pub runtime_id: RuntimeId,
    /// Worker to resume.
    pub worker_id: WorkerId,
}

/// Request to step one debugger-stopped Frame.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct StepRequest {
    /// World containing the stopped Frame.
    pub world_id: WorldId,
    /// Frame to step.
    pub frame_id: FrameId,
    /// Step granularity.
    pub step: Step,
}

/// Request to evaluate one TS++ expression.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct EvaluateRequest {
    /// World containing the selected context.
    pub world_id: WorldId,
    /// Committed Moment to inspect, or the current World when absent.
    pub moment: Option<Moment>,
    /// Evaluation context.
    pub target: EvaluationTarget,
    /// TS++ source expression.
    pub expression: String,
    /// Mutation behavior.
    pub mode: EvaluationMode,
    /// Evaluation resource limits.
    pub limits: EvaluationLimits,
}

/// Request to read captured Frames from one World.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct ReadFramesRequest {
    /// World to inspect.
    pub world_id: WorldId,
    /// Committed Moment to inspect, or the current World when absent.
    pub moment: Option<Moment>,
    /// Worker to inspect, or every Worker when absent.
    pub worker_id: Option<WorkerId>,
}

/// Request to read one World's mapped memory Regions.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct ReadMemoryMapRequest {
    /// World to inspect.
    pub world_id: WorldId,
    /// Committed Moment to inspect, or the current World when absent.
    pub moment: Option<Moment>,
}

/// Request to read one exact World memory range.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct ReadMemoryRequest {
    /// World to inspect.
    pub world_id: WorldId,
    /// Exact committed Moment to inspect.
    pub moment: Moment,
    /// Exact logical memory range to read.
    pub range: MemoryRange,
}

/// Request to list managed heap Allocations.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct ListAllocationsRequest {
    /// World to inspect.
    pub world_id: WorldId,
    /// Exact committed Moment to inspect.
    pub moment: Moment,
    /// Allocation preceding the requested stream.
    pub after: Option<AllocationId>,
}

/// Request to list heap Roots.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct ListRootsRequest {
    /// World to inspect.
    pub world_id: WorldId,
    /// Exact committed Moment to inspect.
    pub moment: Moment,
    /// Root preceding the requested stream.
    pub after: Option<RootId>,
}

/// Request to list outgoing References from one Allocation.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct ListReferencesRequest {
    /// World to inspect.
    pub world_id: WorldId,
    /// Exact committed Moment to inspect.
    pub moment: Moment,
    /// Source Allocation.
    pub allocation_id: AllocationId,
    /// Reference byte offset preceding the requested stream.
    pub after: Option<u64>,
}

/// Request to list the Breakpoints in one World.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct ListBreakpointsRequest {
    /// World to inspect.
    pub world_id: WorldId,
    /// Committed Moment to inspect, or the current World when absent.
    pub moment: Option<Moment>,
}

/// Request to add one Breakpoint.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct AddBreakpointRequest {
    /// World receiving the Breakpoint.
    pub world_id: WorldId,
    /// Executable Program point selection.
    pub filter: PointFilter,
}

/// Request to update one Breakpoint.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct UpdateBreakpointRequest {
    /// World containing the Breakpoint.
    pub world_id: WorldId,
    /// Complete replacement Breakpoint.
    pub breakpoint: Breakpoint,
}

/// Request to remove one Breakpoint.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct RemoveBreakpointRequest {
    /// World containing the Breakpoint.
    pub world_id: WorldId,
    /// Breakpoint to remove.
    pub breakpoint_id: program::BreakpointId,
}

/// Request to list the Watchpoints in one World.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct ListWatchpointsRequest {
    /// World to inspect.
    pub world_id: WorldId,
    /// Committed Moment to inspect, or the current World when absent.
    pub moment: Option<Moment>,
}

/// Request to add one Watchpoint.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct AddWatchpointRequest {
    /// World receiving the Watchpoint.
    pub world_id: WorldId,
    /// Memory access selection.
    pub filter: MemoryFilter,
}

/// Request to update one Watchpoint.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct UpdateWatchpointRequest {
    /// World containing the Watchpoint.
    pub world_id: WorldId,
    /// Complete replacement Watchpoint.
    pub watchpoint: Watchpoint,
}

/// Request to remove one Watchpoint.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct RemoveWatchpointRequest {
    /// World containing the Watchpoint.
    pub world_id: WorldId,
    /// Watchpoint to remove.
    pub watchpoint_id: program::WatchpointId,
}

/// Request to list the Probes in one World.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct ListProbesRequest {
    /// World to inspect.
    pub world_id: WorldId,
    /// Committed Moment to inspect, or the current World when absent.
    pub moment: Option<Moment>,
}

/// Request to add one Probe.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct AddProbeRequest {
    /// World receiving the Probe.
    pub world_id: WorldId,
    /// Runtime activity selection.
    pub filter: ProbeFilter,
    /// Action performed for each match.
    pub action: ProbeAction,
}

/// Request to update one Probe.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct UpdateProbeRequest {
    /// World containing the Probe.
    pub world_id: WorldId,
    /// Complete replacement Probe.
    pub probe: Probe,
}

/// Request to remove one Probe.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct RemoveProbeRequest {
    /// World containing the Probe.
    pub world_id: WorldId,
    /// Probe to remove.
    pub probe_id: ProbeId,
}
